File: benchmarks/Java/safety/simple/even_list.java
class List extends Object {
    List() {
        super();
    }
    Boolean iseven() {
        assert false;
    }
}

class Nil extends List {
    Nil() {
        super();
    }
    Boolean iseven() {
        return true;
    }
}

class Cons extends List {
    Int hd;
    List tl;
    Cons(Int hd, List tl) {
        super();
        this.hd = hd;
        this.tl = tl;
    }
    Boolean iseven() {
        return !this.tl.iseven();
    }
}

main(){
    assert new Cons(1, new Cons(2, new Nil())).iseven();
}

File: benchmarks/OCaml/safety/objects/even_list.ml
class virtual list = object (s)
  method virtual iseven : bool
end

class nil = object
  inherit list
  method iseven = true
end

class cons (hd:int) (tl:list)  = object
  inherit list
  method iseven = not (tl#iseven)
end

let main n =
  assert (new cons n (new cons n new nil))#iseven

[@@@types"
iseven_cons :: (var74 : int) ->
               (var72 : obj) ->
               (var70 : obj) ->
               {var68:bool | iseven_cons[0:1]((var74),(var72),(var70),(var68))}
"]
